crates/ruff_server/src/server/api/requests/format.rs

```diff
@@ -80,10 +92,32 @@ pub(super) fn format_document(snapshot: &DocumentSnapshot) -> Result<super::Form
 fn format_text_document(
     text_document: &TextDocument,
     source_type: PySourceType,
+    file_path: &Path,
+    file_resolver_settings: &FileResolverSettings,
     formatter_settings: &FormatterSettings,
     encoding: PositionEncoding,
     is_notebook: bool,
 ) -> Result<super::FormatResponse> {
+    // If the document is excluded, return early.
+    for path in file_path.ancestors() {
+        if let Some(basename) = path.file_name() {
+            let path = Candidate::new(&path);
+            let basename = Candidate::new(basename);
+            if match_candidate_exclusion(&path, &basename, &file_resolver_settings.exclude) {
+                tracing::debug!("Ignored path via `exclude`: {}", file_path.display());
+                return Ok(None);
+            }
+            if match_candidate_exclusion(&path, &basename, &file_resolver_settings.extend_exclude) {
+                tracing::debug!("Ignored path via `extend-exclude`: {}", file_path.display());
+                return Ok(None);
+            }
+            if match_candidate_exclusion(&path, &basename, &formatter_settings.exclude) {
+                tracing::debug!("Ignored path via `format.exclude`: {}", file_path.display());
+                return Ok(None);
+            }
+        }
+    }
```

Comment:
I intentionally did _not_ add this to the range format action. That feels intentional enough to me that it's fine?

Reply:
I think we might want to add it to range formatting as well. Otherwise, Ruff will always format when auto-format is enabled for modifications in VS Code even if the file is excluded.
```json
{
  "editor.formatOnSave": false,
  "editor.formatOnSaveMode": "modifications"
}
```

Reply:
Ok, sounds good.